Allow pollworkers to accumulate results from multiple print machines

carolinemodic opened the pull request:
Allows users to go 1-by-1 around print-enabled bmd machines (happy meal, or a print kiosk) and accumulate all of the results to one machine. 
https://zube.io/votingworks/vxsuite/c/3105 

Video

https://user-images.githubusercontent.com/14897017/112395399-84d11500-8cbb-11eb-997f-adf0468c826d.mov

Current Outputted report:
[output.pdf](https://github.com/votingworks/vxsuite/files/6200977/output.pdf)


Followups still todo after this PR
- Redesign the machine info in the PollsReport to not look ugly.
- Add undervotes and write in information to the PrecinctTallyReport (not strictly related to this feature but seems like something we should have)

eventualbuddha commented on apps/bmd/src/pages/PollWorkerScreen.tsx:
```diff
@@ -330,29 +466,70 @@ const PollWorkerScreen: React.FC<Props> = ({
           />
         )}
+        {isSavingTally && (
+          <Modal content={<Loading>Saving to pollworker card…</Loading>} />
```
Either just "card" or "Poll Worker card" or "poll worker card" or something like that.
```suggestion
          <Modal content={<Loading>Saving to card…</Loading>} />
```

eventualbuddha commented on apps/bmd/src/components/PollsReport.tsx:
```diff
@@ -75,6 +80,7 @@ interface Props {
   isLiveMode: boolean
   isPollsOpen: boolean
   machineConfig: MachineConfig
+  machineMetadata?: CardTallyMetadataEntry[]
```
```suggestion
  machineMetadata?: readonly CardTallyMetadataEntry[]
```

eventualbuddha commented on apps/bmd/src/config/types.ts:
```diff
@@ -111,6 +111,17 @@ export type Tally = (
+  tally: Tally
+  metadata: CardTallyMetadataEntry[]
+  totalBallotsPrinted: number
+}
```
Use `interface` and `readonly`?

eventualbuddha commented on apps/bmd/src/pages/PollWorkerScreen.test.tsx:
```diff
@@ -116,9 +130,265 @@ test('live mode on election day', async () => {
+  )
+
+  // eslint-disable-next-line no-restricted-syntax
+  expect(queryByText('Combine Results Reports')).toBe(null)
```
If you wanted to avoid the eslint disable directive you could do this:
```suggestion
  expect(queryByText('Combine Results Reports')).toBeNull()
```

eventualbuddha commented on apps/bmd/src/pages/PollWorkerScreen.tsx:
```diff
@@ -330,29 +466,70 @@ const PollWorkerScreen: React.FC<Props> = ({
+                  Do you want to print the combined results report from the{' '}
+                  {pluralize('machine', machineMetadata.length, true)} (
+                  {machineMetadata.map((m) => m.machineId).join(', ')}) and
+                  clear data from the card?
```
This reads slightly better to me:
```suggestion
                  erase the card?
```

benadida replied:
"erase the card" might imply that it's no longer a poll worker card... could sound scarier. Maybe we should be even more specific, "clear tally data from the card"?

carolinemodic replied:
I like clear tally data.

eventualbuddha commented on apps/bmd/src/pages/PollWorkerScreen.tsx:
```diff
@@ -182,6 +219,81 @@ const PollWorkerScreen: React.FC<Props> = ({
+    })
+    setIsSavingTally(false)
+  }
+  const metadataRows = []
```
Give this a type so it isn't inferred as `any[]`.
```suggestion
  const metadataRows: React.ReactChild[] = []
```

eventualbuddha commented on apps/bmd/src/pages/PollWorkerScreen.tsx:
```diff
@@ -330,29 +466,70 @@ const PollWorkerScreen: React.FC<Props> = ({
               election={election}
               isLiveMode={isLiveMode}
-              isPollsOpen={isPollsOpen}
+              isPollsOpen={isPrintingCombinedReport ? false : !isPollsOpen}
```
Why did this change to negate `isPollsOpen`? It reads like a mistake, but the logic is a little hard to follow so it's not clear whether it is or not.

benadida replied:
agreed, I'm confused too.

carolinemodic replied:
So this report is printed right before the value of isPollsOpen is updated, currently it prints the correct report by having the code in PollsReport print a polls closed report when isPollOpen is true and a polls open report when isPollsOpen is false. I thought this was super confusing so I updated PollsReport to intepret isPollsOpen more "correctly" (true means print polls are open) and instead the confusion is moved into here. 

I agree its still confusing.... but I think this is a bit better then having the PollsReport component do something that seems backwords.

benadida replied:
I can live with this, but would recommend a comment explaining the meaning of `isPollsOpen` here.

eventualbuddha commented on apps/bmd/src/utils/Card.ts:
```diff
@@ -159,8 +159,7 @@ export class MemoryCard implements Card {
     const { longValue } = this
-
-    if (!longValue) {
+    if (!longValue || longValue.length === 0) {
```
What triggered this change? Should we add a test for it?

benadida replied:
I suspect it's that we are clearing the card data on the poll worker card... and in the case of the voter card, where this also happens, the case has never come up because an expired voter card doesn't lead to reading the long value.

carolinemodic replied:
I was getting crazy explosions in integration tests before updating this (this is in MemoryCard, so this just impacts the test environment ) after clearing the long value, longValue here would be an empty UInt8Array and when you pass that to the TextDecoder below it exploded.

eventualbuddha commented on apps/bmd/src/utils/tallies.test.ts:
```diff
@@ -192,3 +192,117 @@ test('tallies votes across many contests appropriately', () => {
+  }
+
+  // Modify a yes no contest
+  tally1[12] = {
```
These indexes feel fragile. Can we verify them or compute them?

eventualbuddha commented on apps/bmd/src/utils/tallies.ts:
```diff
@@ -6,10 +8,114 @@ import {
+  tally1: CandidateVoteTally,
+  tally2: CandidateVoteTally
+): CandidateVoteTally => {
+  const candidates = []
```
Give it an annotation so it isn't inferred as `any[]`:
```suggestion
  const candidates: number[] = []
```

eventualbuddha commented on apps/bmd/src/utils/tallies.ts:
```diff
@@ -6,10 +8,114 @@ import {
+  tally2: CandidateVoteTally
+): CandidateVoteTally => {
+  const candidates = []
+  assert(tally1.candidates.length === tally2.candidates.length)
```
This would give you more information on a mismatch:
```suggestion
  assert.equal(tally1.candidates.length, tally2.candidates.length)
```

eventualbuddha commented on apps/bmd/src/utils/tallies.ts:
```diff
@@ -6,10 +8,114 @@ import {
+      (v) => v.name === writeInVote.name
+    )
+    if (existingWriteInIdx >= 0) {
+      writeInVotes[existingWriteInIdx].tally += writeInVote.tally
```
This mutates `tally1`. I don't think it's clear from the function signature that it would do that. Can we make it not mutate its arguments?

eventualbuddha commented on apps/bmd/src/utils/tallies.ts:
```diff
@@ -6,10 +8,114 @@ import {
+): Tally => {
+  assert(
+    election.contests.length === tally1.length &&
+      tally1.length === tally2.length
```
I suggest using two assertions using `assert.equal` to get more information in the event of a crash.

eventualbuddha commented on apps/bmd/src/utils/tallies.ts:
```diff
@@ -6,10 +8,114 @@ import {
+    election.contests.length === tally1.length &&
+      tally1.length === tally2.length
+  )
+  const combinedTally = [] as Tally
```
```suggestion
  const combinedTally: Tally = []
```

carolinemodic commented:
Updated to address Brian's comments, and some small nits raised in show and tell 
- Removes double ellipsis
- Added ballot counts printed per machine to data model and table on the final report

eventualbuddha commented on apps/bmd/src/config/types.ts:
```diff
@@ -111,6 +111,18 @@ export type Tally = (
+
+export interface CardTally {
+  readonly tally: Tally
+  readonly metadata: CardTallyMetadataEntry[]
```
```suggestion
  readonly metadata: readonly CardTallyMetadataEntry[]
```

eventualbuddha commented on apps/bmd/src/utils/tallies.ts:
```diff
@@ -6,10 +8,112 @@ import {
+  for (let i = 0; i < tally1.candidates.length; i++) {
+    candidates.push(tally1.candidates[i] + tally2.candidates[i])
+  }
+  const writeInVotes = [...tally1.writeIns]
```
This copies the array, but not its contents. So the mutation you have inside the loop is still mutating the original.

carolinemodic replied:
I'm actually refactoring how write ins work and remove this code in https://github.com/votingworks/vxsuite/pull/403

benadida replied:
just to be extra safe, I would suggest merging 403 into this first, so we can confirm this mutation no longer happens, before merging this to `main`. I'm being super careful here because we have been bitten before by mutating data structures, and so good to be extra extra careful.

benadida reviewed (changes requested):
I agree with Brian's comments overall, and specifically on being careful with the mutating data structures, even if you're fixing them soon. Let's fix them first (maybe by merging the next PR into here), so `main` never has that behavior, just to be safe.

benadida commented:
test response just to see how email responses work.

On Fri, Mar 26, 2021 at 3:50 PM carolinemodic ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In apps/bmd/src/utils/Card.ts
> <https://github.com/votingworks/vxsuite/pull/396#discussion_r602621065>:
>
> > @@ -159,8 +159,7 @@ export class MemoryCard implements Card {
>     */
>    public async readLongObject<T>(): Promise<Optional<T>> {
>      const { longValue } = this
> -
> -    if (!longValue) {
> +    if (!longValue || longValue.length === 0) {
>
> I was getting crazy explosions in integration tests before updating this
> (this is in MemoryCard, so this just impacts the test environment ) after
> clearing the long value, longValue here would be an empty UInt8Array and
> when you pass that to the TextDecoder below it exploded.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/votingworks/vxsuite/pull/396#discussion_r602621065>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAENCPZ3PZDLIBRJ3OZQMLTFUFR7ANCNFSM4ZYHT2MA>
> .
>

benadida commented on apps/bmd/src/AppRoot.tsx:
```diff
@@ -690,10 +724,16 @@ const AppRoot: React.FC<Props> = ({
         case 'pollworker': {
+          const isValid =
+            cardData.h === optionalElectionDefinition?.electionHash
+          const possibleCardTally: Optional<CardTally> = isValid
+            ? ((await card.readLongObject()) as Optional<CardTally>)
+            : undefined
```
ah so this means that it will wait to declare a poll worker card is present until we read the long value. I can live with that for now, but it will mean that inserting the poll worker card will cause a longer lag before the screen reacts. Let's keep an eye on this when we test on live harware and see how this looks.

One tweak here: no reason to go read the long object if there is no long value stored on the card.

carolinemodic replied:
This seems to also be how we handle reading the long value on voter cards, in practice I haven't really noticed any major lag.

benadida replied:
good point. Maybe an optimization we can make later on both.

benadida commented on apps/bmd/src/pages/PollWorkerScreen.tsx:
```diff
@@ -182,6 +219,83 @@ const PollWorkerScreen: React.FC<Props> = ({
+    metadataRows.push(
+      ...talliesOnCard.metadata
+        .slice()
+        .sort((a, b) => b.timeSaved - a.timeSaved)
```
I believe this is a safe `sort` because of the `...` expansion, right? /cc @eventualbuddha

eventualbuddha replied:
It's safe because of the `.slice()`, which makes a copy. Another way would have been `[...talliesOnCard.metadata].sort()`, which is probably what you're thinking of @benadida.

benadida reviewed (approved):
Couple of small things to confirm / look at, otherwise good for me.

eventualbuddha commented on apps/bmd/src/AppRoot.tsx:
```diff
@@ -207,7 +210,11 @@ const initialAppState: Readonly<State> = {
+  | {
+      type: 'processPollWorkerCard'
+      isPollWorkerCardValid: boolean
+      talliesOnCard: Optional<CardTally>
```
This would be more idiomatic:
```suggestion
      talliesOnCard?: CardTally
```

eventualbuddha commented on apps/bmd/src/AppRoot.tsx:
```diff
@@ -231,6 +238,10 @@ type AppAction =
   | { type: 'maintainCardlessBallot' }
+  | {
+      type: 'updatePollWorkerCardTally'
+      talliesOnCard: Optional<CardTally>
```
Ditto:

```suggestion
      talliesOnCard?: CardTally
```

eventualbuddha commented on apps/bmd/src/AppRoot.tsx:
```diff
@@ -690,10 +724,20 @@ const AppRoot: React.FC<Props> = ({
+          const isValid =
+            cardData.h === optionalElectionDefinition?.electionHash
+          /* istanbul ignore next */
+          // const shouldFetchLong = !!(isValid && longValueExists)
```
Delete?

carolinemodic replied:
lol oops

eventualbuddha commented on apps/bmd/src/pages/PollWorkerScreen.test.tsx:
```diff
@@ -116,9 +130,289 @@ test('live mode on election day', async () => {
+    />
+  )
+
+  // eslint-disable-next-line no-restricted-syntax
```
Comment outdated.
```suggestion
```

eventualbuddha commented on apps/bmd/src/utils/tallies.ts:
```diff
@@ -6,10 +8,104 @@ import {
+    candidates.push(tally1.candidates[i] + tally2.candidates[i])
+  }
+  return {
+    candidates,
```
Someday:
```suggestion
    candidates: zip(tally1.candidates, tally2.candidates, (c1, c2) => c1 + c2),
```
